commit f8328abb878207cbb5e5a2159a07341fa707365f
Author: Jiri Olsa <jolsa@kernel.org>
Date:   Mon Nov 12 14:00:12 2018 +0100

    perf tools: Fix crash on synthesizing the unit
    
    [ Upstream commit fb50c09e923870a358d68b0d58891bd145b8d7c7 ]
    
    Adam reported a record command crash for simple session like:
    
      $ perf record -e cpu-clock ls
    
    with following backtrace:
    
      Program received signal SIGSEGV, Segmentation fault.
      3543            ev = event_update_event__new(size + 1, PERF_EVENT_UPDATE__UNIT, evsel->id[0]);
      (gdb) bt
      #0  perf_event__synthesize_event_update_unit
      #1  0x000000000051e469 in perf_event__synthesize_extra_attr
      #2  0x00000000004445cb in record__synthesize
      #3  0x0000000000444bc5 in __cmd_record
      ...
    
    We synthesize an update event that needs to touch the evsel id array,
    which is not defined at that time. Fix this by forcing the id allocation
    for events with their unit defined.
    
    Reflecting possible read_format ID bit in the attr tests.
    
    Reported-by: Yongxin Liu <yongxin.liu@outlook.com>
    Signed-off-by: Jiri Olsa <jolsa@kernel.org>
    Cc: Adam Lee <leeadamrobert@gmail.com>
    Cc: Alexander Shishkin <alexander.shishkin@linux.intel.com>
    Cc: Namhyung Kim <namhyung@kernel.org>
    Cc: Peter Zijlstra <peterz@infradead.org>
    Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=201477
    Fixes: bfd8f72c2778 ("perf record: Synthesize unit/scale/... in event update")
    Link: http://lkml.kernel.org/r/20181112130012.5424-1-jolsa@kernel.org
    Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
    Signed-off-by: Sasha Levin <sashal@kernel.org>

diff --git a/tools/perf/tests/attr/base-record b/tools/perf/tests/attr/base-record
index 37940665f736..efd0157b9d22 100644
--- a/tools/perf/tests/attr/base-record
+++ b/tools/perf/tests/attr/base-record
@@ -9,7 +9,7 @@ size=112
 config=0
 sample_period=*
 sample_type=263
-read_format=0
+read_format=0|4
 disabled=1
 inherit=1
 pinned=0
diff --git a/tools/perf/util/evsel.c b/tools/perf/util/evsel.c
index 03a72310315f..e7dbdcc8d465 100644
--- a/tools/perf/util/evsel.c
+++ b/tools/perf/util/evsel.c
@@ -1088,7 +1088,7 @@ void perf_evsel__config(struct perf_evsel *evsel, struct record_opts *opts,
 		attr->exclude_user   = 1;
 	}
 
-	if (evsel->own_cpus)
+	if (evsel->own_cpus || evsel->unit)
 		evsel->attr.read_format |= PERF_FORMAT_ID;
 
 	/*
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1040_f8328abb878207cbb5e5a2159a07341fa707365f.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1040_f8328abb878207cbb5e5a2159a07341fa707365f.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
